internal/reconciler/garbagecollector/reconciler.go

```diff
@@ -6,29 +6,118 @@ package garbagecollector
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	configv1alpha1 "github.com/gardener/garden-shoot-trust-configurator/pkg/apis/config/v1alpha1"
+	constants "github.com/gardener/garden-shoot-trust-configurator/pkg/apis/constants"
 )
 
 // Reconciler performs garbage collection.
 type Reconciler struct {
 	Client client.Client
 	Config configv1alpha1.GarbageCollectorControllerConfig
+	Clock  clock.Clock
 }
 
 // Reconcile performs the main reconciliation logic.
 func (r *Reconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
 	log := logf.FromContext(ctx)
 
+	ctx, cancel := controllerutils.GetMainReconciliationContext(ctx, r.Config.SyncPeriod.Duration)
+	defer cancel()
+
 	log.Info("Starting garbage collection")
-	defer log.Info("Garbage collection finished")
 
-	return reconcile.Result{
-		Requeue:      true,
-		RequeueAfter: r.Config.SyncPeriod.Duration,
+	var (
+		label                   = client.MatchingLabels{constants.LabelManagedByKey: constants.LabelManagedByValue}
+		objectsToGarbageCollect = sets.New[string]()
+	)
+
+	objList := &metav1.PartialObjectMetadataList{}
+	objList.SetGroupVersionKind(authenticationv1alpha1.GroupVersion.WithKind("OpenIDConnectList"))
+	if err := r.Client.List(ctx, objList, label); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed listing OIDCs: %w", err)
+	}
+
+	for _, obj := range objList.Items {
+		if obj.CreationTimestamp.Add(r.Config.MinimumObjectLifetime.Duration).UTC().After(r.Clock.Now().UTC()) {
+			// Do not consider recently created objects for garbage collection.
+			continue
+		}
+		objectsToGarbageCollect.Insert(obj.Name)
+	}
+
+	for oidcName := range objectsToGarbageCollect {
+		oidc := authenticationv1alpha1.OpenIDConnect{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: oidcName}, &oidc); err != nil {
+			if apierrors.IsNotFound(err) {
+				// Object already deleted, nothing to do.
+				continue
+			}
+			log.Error(err, "Error retrieving OIDC resource from store", "oidc", oidcName)
+			continue
+		}
+
+		shootNamespacedName, err := parseOIDCResourceName(&oidc)
+		if err != nil {
+			log.Error(err, "Skipping OIDC resource as it has an invalid name", "oidc", oidc.Name)
+			continue
+		}
+
+		shoot := &gardencorev1beta1.Shoot{}
+		err = r.Client.Get(ctx, shootNamespacedName, shoot)
+		if err != nil {
+			if client.IgnoreNotFound(err) != nil {
+				log.Error(err, "Error retrieving shoot from store", "shoot", shootNamespacedName)
+				continue
+			}
+
+			log.Info("Shoot not found, deleting OIDC resource", "shoot", shootNamespacedName, "oidc", oidc.Name)
+			if err := r.Client.Delete(ctx, &oidc); err != nil {
```

Comment:
Should we ignore not found errors?
```suggestion
			if err := client.IgnoreNotFound(r.Client.Delete(ctx, &oidc)); err != nil {
```